File: code/src/wang/relish/algorithm/handler/Handler.java
package wang.relish.algorithm.handler;

public abstract class Handler {

    private Message mMessage;
    private Looper mLooper;


    public Handler() {
        Looper looper = Looper.looperThreadLocal.get();
        if (looper == null) {
            throw new Error("looper not exist where current thread !");
        }
        this.mLooper = looper;
    }

    public Handler(Looper looper) {
        if (looper == null) {
            throw new NullPointerException("");
        }
        this.mLooper = looper;
    }

    public void post(Runnable runnable) {
        this.mMessage = Message.obtain();
        mMessage.callback = runnable;
        mMessage.when = 0L;
        mMessage.what = null;
        this.sendMessage(mMessage);
    }

    public void sendMessage(Message message) {
        message.sendWhen = System.currentTimeMillis();
        message.target = this;
        this.mLooper.getMessageQueue().offer(message);
    }

    protected abstract void handleMessage(Message message);

    public Looper getLooper() {
        return mLooper;
    }

    public void setLooper(Looper mLooper) {
        this.mLooper = mLooper;
    }
}

File: code/src/wang/relish/algorithm/handler/mini/Handler.kt
package wang.relish.algorithm.handler.mini


open class Handler {

    val mLooper: Looper
    val mQueue: MessageQueue

    constructor() {
        mLooper = Looper.myLooper()
        mQueue = mLooper.mQueue
    }

    constructor(looper: Looper) {
        mLooper = looper
        mQueue = mLooper.mQueue
    }

    fun sendMessage(msg: Message) {
        sendMessageDelay(msg, 0L)
    }

    fun sendMessageDelay(msg: Message, upTime: Long) {
        val t = maxOf(upTime, 0L)
        msg.target = this
        mQueue?.enqueueMessage(msg, System.currentTimeMillis() + t)
    }

    open fun handleMessage(msg: Message) {
    }

    fun dispatchMessage(msg: Message) {
        handleMessage(msg)
    }
}